src/mpid/ch4/generic/am/mpidig_am_recv.h

```diff
@@ -240,61 +250,22 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_do_irecv(void *buf, MPI_Aint count, MPI_Data
+                 * a request. Here we simply return `unexp_req`, which is already completed. */
+                *request = unexp_req;
             } else {
-                /* if the unexpected recv is not ready, we put user recv buffer info and let
-                 * transport layer to start the recv. */
-                /* the count for unexpected long message is the data size */
-                MPI_Aint data_sz = MPIDIG_REQUEST(unexp_req, count);
-                /* Matching receive is now posted, tell the netmod/shmmod */
-                MPIR_Datatype_add_ref_if_not_builtin(datatype);
-                MPIDIG_REQUEST(unexp_req, datatype) = datatype;
-                MPIDIG_REQUEST(unexp_req, buffer) = (char *) buf;
-                MPIDIG_REQUEST(unexp_req, count) = count;
-                MPIDIG_REQUEST(unexp_req, recv_ready) = true;
-                if (*request == NULL) {
-                    /* Regular (non-enqueuing) path: MPIDIG is responsbile for allocating
-                     * a request. Here we simply return `unexp_req` */
-                    *request = unexp_req;
-                    /* Mark `match_req` as NULL so that we know nothing else to complete when
-                     * `unexp_req` finally completes. (See below) */
-                    MPIDIG_REQUEST(unexp_req, req->rreq.match_req) = NULL;
-                } else {
-                    /* Enqueuing path: CH4 already allocated a request.
-                     * Record the passed `*request` to `match_req` so that we can complete it
-                     * later when `unexp_req` completes.
-                     * See MPIDI_recv_target_cmpl_cb for actual completion handler. */
-                    MPIDIG_REQUEST(unexp_req, req->rreq.match_req) = *request;
-                }
-                MPIDIG_REQUEST(unexp_req, req->status) &= ~MPIDIG_REQ_UNEXPECTED;
-                MPIDIG_REQUEST(unexp_req, req->seq_no) =
-                    MPL_atomic_fetch_add_uint64(&MPIDI_global.nxt_seq_no, 1);
-                MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
-                                (MPL_DBG_FDEST, "seq_no: me=%" PRIu64 " exp=%" PRIu64,
-                                 MPIDIG_REQUEST(unexp_req, req->seq_no),
-                                 MPL_atomic_load_uint64(&MPIDI_global.exp_seq_no)));
-                MPIDIG_recv_type_init(data_sz, unexp_req);
-                MPIDIG_do_am_recv(unexp_req);
-                MPIR_ERR_CHECK(mpi_errno);
+                /* Enqueuing path: CH4 already allocated request as `*request`.
+                 * Since the real operations has completed in `unexp_req`, here we
+                 * simply copy the status to `*request` and complete it. */
+                (*request)->status = unexp_req->status;
+                MPIR_Request_add_ref(*request);
+                MPID_Request_complete(*request);
+                /* Need to free here because we don't return this to user */
+                MPIDI_CH4_REQUEST_FREE(unexp_req);
```

Comment:
Again, need to handle the rndv case which it is not completed yet. I think the existing mechanism is to attach the pre-allocated request as `matched_request` to the `unexp_req` and let the `target_cmpl_cb` take care of it.

Reply:
No, the matched request is an addition of the WorkQ. I kind of hate the current status of unexp handling and completion. Cleaning those up right now.